design/kanister-prometheus-integration.md

```diff
@@ -0,0 +1,286 @@
+# Enhancing observability in Kanister
+
+<!-- toc -->
+- [Motivation](#motivation)
+- [Scope](#scope)
+- [Detailed Design](#detailed-design)
+  - [Architecture](#architecture)
+    - [Text description](#text-description)
+  - [Low Level APIs](#low-level-apis)
+    - [Metrics Package](#metrics-package)
+      - [Example Initialization Steps for a new CounterVec metric](#example-initialization-steps-for-a-new-countervec-metric)
+    - [Consumer Package](#consumer-package)
+- [Testing](#testing)
+<!-- /toc -->
+
+This document proposes the changes required within Kanister to integrate
+[Prometheus](https://github.com/prometheus/prometheus) and enhance the observability 
+of action sets. 
+
+## Motivation
+Kanister controller already has a registered metrics endpoint `/metrics`. There are no metrics exported other than the default Prometheus metrics that the default handler provides. Adding metrics to track the 
+ActionSets and Blueprints workflow will help improve the overall observability. 
+
+To achieve this, we need to build a framework for exporting metrics from the Kanister controller, and to 
+start with, export some metrics to Prometheus. 
+
+Some example metrics include:
+ActionSets succeeded
+ActionSets failed
+Action duration
+Phase duration, etc.
+
+
+## Scope
+1. Design a framework that allows us to export new
+Kanister metrics to Prometheus easily. 
+2. Add a few fundamental metrics related to ActionSets and Blueprints to start with.
+
+
+## Detailed Design
+
+### Architecture
+
+![Alt text](Metrics_design.png?raw=true "Prometheus Integration Design")
+
+#### Text description
+1. The initializer of the consumer package calls new_metrics, a helper method that talks to Kanister’s metrics package. The result is a new metrics struct that owns all the prometheus collectors.
+
+2. In order to initialize all the required prometheus collectors, the new_metrics method calls the InitCounterVec, InitGaugeVec, InitHistogramVec in the metrics package. It passes the metric names and the specific label names and label values, aka, BoundedLabels, to the metrics package. Once it initializes all the prometheus collectors successfully, it returns an struct that wraps all the collectors, which the consumer package can then use. 
+
+3. The metrics package internally attempts to initialize the collectors and register the specific collectors with prometheus. If the registration fails because the specific metric with label header already exists, the collector will simply be returned to the caller. If the registration fails due to other reasons, then the metrics package will cause a panic, signalling programmer error. In case of the CounterVec, the InitCounterVec attempts to generate all possible permutations of label values and sets each counter 
+within the CounterVec to 0. 
+
+4. Once the collector is created in the metrics package, it will be returned to the consumer package’s new_metrics helper method.
```

Comment:
```suggestion
4. Once the Prometheus metric is created in the metrics package, it will be returned to the consumer package’s new_metrics helper method.
```

Reply:
Fixed